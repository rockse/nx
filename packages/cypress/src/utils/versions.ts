export const nxVersion = require('../../package.json').version;
export const eslintPluginCypressVersion = '^2.10.3';
export const typesNodeVersion = '16.11.7';
export const cypressVersion = '^10.1.0';
export const cypressWebpackVersion = '^2.0.0';
export const webpackHttpPluginVersion = '^5.5.0';
