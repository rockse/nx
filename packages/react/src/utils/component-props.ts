import * as ts from 'typescript';
import { getComponentPropsInterface } from './ast-utils';

// TODO: candidate to refactor with the angular component story
export function getArgsDefaultValue(property: ts.SyntaxKind): string {
  const typeNameToDefault: Record<number, any> = {
    [ts.SyntaxKind.StringKeyword]: "''",
    [ts.SyntaxKind.NumberKeyword]: 0,
    [ts.SyntaxKind.BooleanKeyword]: false,
  };

  const resolvedValue = typeNameToDefault[property];
  if (typeof resolvedValue === undefined) {
    return "''";
  } else {
    return resolvedValue;
  }
}

export function getDefaultsForComponent(
  sourceFile: ts.SourceFile,
  cmpDeclaration: ts.Node
) {
  const propsInterface = getComponentPropsInterface(sourceFile, cmpDeclaration);

  let propsTypeName: string = null;
  let props: {
    name: string;
    defaultValue: any;
  }[] = [];
  let argTypes: {
    name: string;
    type: string;
    actionText: string;
  }[] = [];

  if (propsInterface) {
    propsTypeName = propsInterface.name.text;
    props = propsInterface.members.map((member: ts.PropertySignature) => {
      if (member.type.kind === ts.SyntaxKind.FunctionType) {
        argTypes.push({
          name: (member.name as ts.Identifier).text,
          type: 'action',
          actionText: `${(member.name as ts.Identifier).text} executed!`,
        });
      } else {
        return {
          name: (member.name as ts.Identifier).text,
          defaultValue: getArgsDefaultValue(member.type.kind),
        };
      }
    });
    props = props.filter((p) => p && p.defaultValue !== undefined);
  }
  return { propsTypeName, props, argTypes };
}

export function getImportForType(sourceFile: ts.SourceFile, typeName: string) {
  return sourceFile.statements.find(
    (statement: ts.Node) =>
      ts.isImportDeclaration(statement) &&
      statement.getText().includes(typeName)
  );
}
