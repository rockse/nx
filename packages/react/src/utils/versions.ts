export const nxVersion = require('../../package.json').version;

export const reactVersion = '18.2.0';
export const reactDomVersion = '18.2.0';
export const reactIsVersion = '18.2.0';
export const typesReactVersion = '18.0.13';
export const typesReactDomVersion = '18.0.5';
export const typesReactIsVersion = '17.0.3';

export const typesNodeVersion = '18.0.0';

export const styledComponentsVersion = '5.3.5';
export const typesStyledComponentsVersion = '5.1.25';

export const emotionStyledVersion = '11.9.3';
export const emotionReactVersion = '11.9.3';
export const emotionBabelPlugin = '11.9.2';

// WARNING: This needs to be in sync with Next.js' dependency or else there might be issues.
export const styledJsxVersion = '5.0.2';

export const reactRouterDomVersion = '6.3.0';
export const typesReactRouterDomVersion = '5.3.3';

export const testingLibraryReactVersion = '13.3.0';

export const reduxjsToolkitVersion = '1.8.2';
export const reactReduxVersion = '8.0.2';
export const reactTestRendererVersion = '18.2.0';

export const eslintPluginImportVersion = '2.26.0';
export const eslintPluginJsxA11yVersion = '6.5.1';
export const eslintPluginReactVersion = '7.30.0';
export const eslintPluginReactHooksVersion = '4.6.0';

export const babelPluginStyledComponentsVersion = '1.10.7';

// used for cypress component tests
export const cypressReactVersion = '^6.0.0';
