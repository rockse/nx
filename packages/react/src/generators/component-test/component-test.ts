import { assertMinimumCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import {
  generateFiles,
  joinPathFragments,
  readProjectConfiguration,
  Tree,
} from '@nrwl/devkit';
import { basename, dirname, extname, relative } from 'path';
import * as ts from 'typescript';
import {
  findExportDeclarationsForJsx,
  getComponentNode,
} from '../../utils/ast-utils';
import { getDefaultsForComponent } from '../../utils/component-props';
import { ComponentTestSchema } from './schema';

export function componentTestGenerator(
  tree: Tree,
  options: ComponentTestSchema
) {
  assertMinimumCypressVersion(10);

  const projectConfig = readProjectConfiguration(tree, options.project);

  const normalizedPath = options.componentPath.startsWith(
    projectConfig.sourceRoot
  )
    ? relative(projectConfig.sourceRoot, options.componentPath)
    : options.componentPath;

  const componentPath = joinPathFragments(
    projectConfig.sourceRoot,
    normalizedPath
  );

  if (tree.exists(componentPath)) {
    generateSpecsForComponents(tree, componentPath);
  }
}

function generateSpecsForComponents(tree: Tree, filePath: string) {
  const sourceFile = ts.createSourceFile(
    filePath,
    tree.read(filePath, 'utf-8'),
    ts.ScriptTarget.Latest,
    true
  );

  const cmpNodes = findExportDeclarationsForJsx(sourceFile);
  const componentDir = dirname(filePath);
  const ext = extname(filePath);
  const fileName = basename(filePath, ext);
  const defaultExport = getComponentNode(sourceFile);

  if (cmpNodes?.length) {
    const components = cmpNodes.map((cmp) => {
      const defaults = getDefaultsForComponent(sourceFile, cmp);
      const isDefaultExport = defaultExport
        ? (defaultExport as any).name.text === (cmp as any).name.text
        : false;
      return {
        isDefaultExport,
        props: [...defaults.props, ...defaults.argTypes],
        name: (cmp as any).name.text as string,
        typeName: defaults.propsTypeName,
      };
    });
    const namedImports = components
      .reduce((imports, cmp) => {
        if (cmp.typeName) {
          imports.push(cmp.typeName);
        }
        if (cmp.isDefaultExport) {
          return imports;
        }

        imports.push(cmp.name);
        return imports;
      }, [])
      .join(', ');

    const namedImportStatement =
      namedImports.length > 0 ? `, { ${namedImports} }` : '';

    generateFiles(tree, joinPathFragments(__dirname, 'files'), componentDir, {
      fileName,
      components,
      importStatement: defaultExport
        ? `import ${
            (defaultExport as any).name.text
          }${namedImportStatement} from './${fileName}'`
        : `import { ${namedImports} } from './${fileName}'`,
      ext,
    });
  }
}

export default componentTestGenerator;
