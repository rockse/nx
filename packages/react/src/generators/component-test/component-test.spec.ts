import { assertMinimumCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import { Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import libraryGenerator from '../library/library';
import { componentTestGenerator } from './component-test';

jest.mock('@nrwl/cypress/src/utils/cypress-version');
describe(componentTestGenerator.name, () => {
  let tree: Tree;
  let mockedAssertMinimumCypressVersion: jest.Mock<
    ReturnType<typeof assertMinimumCypressVersion>
  > = assertMinimumCypressVersion as never;
  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });
  it('should create component test for tsx files', async () => {
    mockedAssertMinimumCypressVersion.mockReturnValue();
    await libraryGenerator(tree, {
      linter: Linter.EsLint,
      name: 'some-lib',
      skipFormat: true,
      skipTsConfig: false,
      style: 'scss',
      unitTestRunner: 'none',
      component: true,
    });

    componentTestGenerator(tree, {
      project: 'some-lib',
      componentPath: 'lib/some-lib.tsx',
    });

    expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
  });

  it('should create component test for js files', async () => {
    mockedAssertMinimumCypressVersion.mockReturnValue();
    await libraryGenerator(tree, {
      linter: Linter.EsLint,
      name: 'some-lib',
      skipFormat: true,
      skipTsConfig: false,
      style: 'scss',
      unitTestRunner: 'none',
      component: true,
      js: true,
    });

    componentTestGenerator(tree, {
      project: 'some-lib',
      componentPath: 'lib/some-lib.js',
    });

    expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.js')).toBeTruthy();
  });

  it('should not throw if path is invalid', async () => {
    mockedAssertMinimumCypressVersion.mockReturnValue();
    await libraryGenerator(tree, {
      linter: Linter.EsLint,
      name: 'some-lib',
      skipFormat: true,
      skipTsConfig: false,
      style: 'scss',
      unitTestRunner: 'none',
      component: true,
    });

    expect(() => {
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'lib/blah/abc-123.blah',
      });
    }).not.toThrow();
  });

  it('should handle being provided the full path to the component', async () => {
    mockedAssertMinimumCypressVersion.mockReturnValue();
    await libraryGenerator(tree, {
      linter: Linter.EsLint,
      name: 'some-lib',
      skipFormat: true,
      skipTsConfig: false,
      style: 'scss',
      unitTestRunner: 'none',
      component: true,
    });

    componentTestGenerator(tree, {
      project: 'some-lib',
      componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
    });

    expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
  });

  describe('multiple components per file', () => {
    it('should handle props', async () => {
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });

      tree.write(
        'libs/some-lib/src/lib/some-lib.tsx',
        `
${tree.read('libs/some-lib/src/lib/some-lib.tsx')}

/* eslint-disable-next-line */
export interface AnotherCmpProps {
  handleClick: () => void;
  text: string;
  count: number;
  isOkay: boolean;
}

export function AnotherCmp(props: AnotherCmpProps) {
 return <button onClick="{handleClick}">{props.text}</button>;
}
`
      );
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });
    it('should handle no props', async () => {
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });

      tree.write(
        'libs/some-lib/src/lib/some-lib.tsx',
        `
${tree.read('libs/some-lib/src/lib/some-lib.tsx')}

export function AnotherCmp() {
 return <button>AnotherCmp</button>;
}
`
      );
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });
    it('should handle default export', async () => {
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });

      tree.write(
        'libs/some-lib/src/lib/some-lib.tsx',
        `
/* eslint-disable-next-line */
export interface AnotherCmpProps {
  handleClick: () => void;
  text: string;
  count: number;
  isOkay: boolean;
}

export default function AnotherCmp(props: AnotherCmpProps) {
 return <button onClick="{handleClick}">{props.text}</button>;
}

export function AnotherCmp2() {
 return <button>AnotherCmp</button>;
}
`
      );
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });

    it('should handle named exports', async () => {
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });

      tree.write(
        'libs/some-lib/src/lib/some-lib.tsx',
        `
/* eslint-disable-next-line */
export interface AnotherCmpProps {
  handleClick: () => void;
  text: string;
  count: number;
  isOkay: boolean;
}

export function AnotherCmp(props: AnotherCmpProps) {
 return <button onClick="{handleClick}">{props.text}</button>;
}

export function AnotherCmp2() {
 return <button>AnotherCmp2</button>;
}
`
      );
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });
  });

  describe('single component per file', () => {
    it('should handle props', async () => {
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });

      tree.write(
        'libs/some-lib/src/lib/some-lib.tsx',
        `
/* eslint-disable-next-line */
export interface AnotherCmpProps {
  handleClick: () => void;
  text: string;
  count: number;
  isOkay: boolean;
}

export function AnotherCmp(props: AnotherCmpProps) {
 return <button onClick="{handleClick}">{props.text}</button>;
}
`
      );
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });
    it('should handle no props', async () => {
      // this is the default behavior of the library component generator
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });
    it('should handle default export', async () => {
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });

      tree.write(
        'libs/some-lib/src/lib/some-lib.tsx',
        `
/* eslint-disable-next-line */
export interface AnotherCmpProps {
  handleClick: () => void;
  text: string;
  count: number;
  isOkay: boolean;
}

export default function AnotherCmp(props: AnotherCmpProps) {
 return <button onClick="{handleClick}">{props.text}</button>;
}
`
      );
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });
    it('should handle named exports', async () => {
      mockedAssertMinimumCypressVersion.mockReturnValue();
      await libraryGenerator(tree, {
        linter: Linter.EsLint,
        name: 'some-lib',
        skipFormat: true,
        skipTsConfig: false,
        style: 'scss',
        unitTestRunner: 'none',
        component: true,
      });

      tree.write(
        'libs/some-lib/src/lib/some-lib.tsx',
        `
/* eslint-disable-next-line */
export interface AnotherCmpProps {
  handleClick: () => void;
  text: string;
  count: number;
  isOkay: boolean;
}

export function AnotherCmp(props: AnotherCmpProps) {
 return <button onClick="{handleClick}">{props.text}</button>;
}
`
      );
      componentTestGenerator(tree, {
        project: 'some-lib',
        componentPath: 'libs/some-lib/src/lib/some-lib.tsx',
      });

      expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
      expect(
        tree.read('libs/some-lib/src/lib/some-lib.cy.tsx', 'utf-8')
      ).toMatchSnapshot();
    });
  });
});
