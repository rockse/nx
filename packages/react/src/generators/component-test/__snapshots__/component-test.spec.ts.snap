// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componentTestGenerator multiple components per file should handle default export 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import AnotherCmp, { AnotherCmpProps, AnotherCmp2 } from './some-lib'


describe(AnotherCmp.name, () => {
  let props: AnotherCmpProps;

  beforeEach(() => {
    props = {
      text: '',
      count: 0,
      isOkay: false,
      handleClick: undefined
    }
  })

  it('renders', () => {
    mount(<AnotherCmp {...props}/>)
  })
})

describe(AnotherCmp2.name, () => {
  it('renders', () => {
    mount(<AnotherCmp2 />)
  })
})

"
`;

exports[`componentTestGenerator multiple components per file should handle named exports 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import { AnotherCmpProps, AnotherCmp, AnotherCmp2 } from './some-lib'


describe(AnotherCmp.name, () => {
  let props: AnotherCmpProps;

  beforeEach(() => {
    props = {
      text: '',
      count: 0,
      isOkay: false,
      handleClick: undefined
    }
  })

  it('renders', () => {
    mount(<AnotherCmp {...props}/>)
  })
})

describe(AnotherCmp2.name, () => {
  it('renders', () => {
    mount(<AnotherCmp2 />)
  })
})

"
`;

exports[`componentTestGenerator multiple components per file should handle no props 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import SomeLib, { SomeLibProps, AnotherCmp } from './some-lib'


describe(SomeLib.name, () => {
  let props: SomeLibProps;

  beforeEach(() => {
    props = {
    }
  })

  it('renders', () => {
    mount(<SomeLib {...props}/>)
  })
})

describe(AnotherCmp.name, () => {
  it('renders', () => {
    mount(<AnotherCmp />)
  })
})

"
`;

exports[`componentTestGenerator multiple components per file should handle props 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import SomeLib, { SomeLibProps, AnotherCmpProps, AnotherCmp } from './some-lib'


describe(SomeLib.name, () => {
  let props: SomeLibProps;

  beforeEach(() => {
    props = {
    }
  })

  it('renders', () => {
    mount(<SomeLib {...props}/>)
  })
})

describe(AnotherCmp.name, () => {
  let props: AnotherCmpProps;

  beforeEach(() => {
    props = {
      text: '',
      count: 0,
      isOkay: false,
      handleClick: undefined
    }
  })

  it('renders', () => {
    mount(<AnotherCmp {...props}/>)
  })
})

"
`;

exports[`componentTestGenerator single component per file should handle default export 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import AnotherCmp, { AnotherCmpProps } from './some-lib'


describe(AnotherCmp.name, () => {
  let props: AnotherCmpProps;

  beforeEach(() => {
    props = {
      text: '',
      count: 0,
      isOkay: false,
      handleClick: undefined
    }
  })

  it('renders', () => {
    mount(<AnotherCmp {...props}/>)
  })
})

"
`;

exports[`componentTestGenerator single component per file should handle named exports 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import { AnotherCmpProps, AnotherCmp } from './some-lib'


describe(AnotherCmp.name, () => {
  let props: AnotherCmpProps;

  beforeEach(() => {
    props = {
      text: '',
      count: 0,
      isOkay: false,
      handleClick: undefined
    }
  })

  it('renders', () => {
    mount(<AnotherCmp {...props}/>)
  })
})

"
`;

exports[`componentTestGenerator single component per file should handle no props 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import SomeLib, { SomeLibProps } from './some-lib'


describe(SomeLib.name, () => {
  let props: SomeLibProps;

  beforeEach(() => {
    props = {
    }
  })

  it('renders', () => {
    mount(<SomeLib {...props}/>)
  })
})

"
`;

exports[`componentTestGenerator single component per file should handle props 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import { AnotherCmpProps, AnotherCmp } from './some-lib'


describe(AnotherCmp.name, () => {
  let props: AnotherCmpProps;

  beforeEach(() => {
    props = {
      text: '',
      count: 0,
      isOkay: false,
      handleClick: undefined
    }
  })

  it('renders', () => {
    mount(<AnotherCmp {...props}/>)
  })
})

"
`;
